lib/elixir/pages/Library Guidelines.md

```diff
@@ -0,0 +1,242 @@
+
+Projects are often made available to other developers [by publishing a Hex package](https://hex.pm/docs/publish). Hex also [supports private packages for organizations](https://hex.pm/pricing). If ExDoc is configured for the Mix project, publishing a package on Hex will also automatically publish the generated documentation to [HexDocs](https://hexdocs.pm).
+
+## Anti-patterns
+
+In this section we document common anti-patterns to avoid when writing libraries.
+
+### Avoid using exceptions for control-flow
+
+You should avoid using exceptions for control-flow. For example, instead of:
+
+```elixir
+try do
+  contents = File.read!("some_path_that_may_or_may_not_exist")
+  {:it_worked, contents}
+rescue
+  File.Error ->
+    :it_failed
+end
+```
+
+you should prefer:
+
+```elixir
+case File.read("some_path_that_may_or_may_not_exist") do
+  {:ok, contents} -> {:it_worked, contents}
+  {:error, _} -> :it_failed
+end
+```
+
+As a library author, it is your responsibility to make sure users are not required to use exceptions for control-flow in their applications. You can follow the same convention as Elixir here, using the name without `!` for returning `:ok`/`:error` tuples and appending `!` for a version of the function which raises an exception.
+
+It is important to note that a name without `!` does not mean a function will never raise. For example, even `File.read/1` can fail in case of bad arguments:
+
+```iex
+iex> File.read(1)
+** (FunctionClauseError) no function clause matching in IO.chardata_to_string/1
+```
+
+The usage of `:ok`/`:error` tuples is about the domain that the function works on, in this case, filesystem access. Bad arguments, logical errors, invalid options should raise regardless of the function name. If in doubt, prefer to return tuples instead of raising, as users of your library can always match on the results and raise if necessary.
+
+### Avoid working with invalid data
+
+Elixir programs should prefer to validate data as close to the end user as possible, so the errors are easy to locate and fix. This practice also saves you from writing defensive code in the internals of the library.
+
+For example, imagine you have an API that receives a filename as a binary. At some point you will want to write to this file. You could have a function like this:
+
+```elixir
+def my_fun(some_arg, file_to_write_to, options \\ []) do
+  ...some code...
+  AnotherModuleInLib.invoke_something_that_will_eventually_write_to_file(file_to_write_to)
+  ...more code...
+end
+```
+
+The problem with the code above is that, if the user supplies an invalid input, the error will be raised deep inside the library, which makes it confusing for users. Furthermore, when you don't validate the values at the boundary, the internals of your library are never quite sure which kind of values they are working with.
+
+A better function definition would be:
+
+```elixir
+def my_fun(some_arg, file_to_write_to, options \\ []) do
```

Comment:
It seems this function definition is the same as the one above, does it miss the guard?

Reply:
Oh, the guard is missing. :)